File: src/main/java/com/ptaf/ui/hooks/Hooks.java
package com.ptaf.ui.hooks;

import com.microsoft.playwright.Browser;
import com.microsoft.playwright.BrowserContext;
import com.microsoft.playwright.Page;
import com.ptaf.api.handlers.ApiRequestHandler;
import com.ptaf.db.handlers.DatabaseHandler;
import com.ptaf.ui.pages.PageCommonMethods;
import com.ptaf.ui.utils.ConfigurationProperties;
import com.ptaf.ui.utils.BrowserFactory;
import io.cucumber.java.After;
import io.cucumber.java.Before;
import io.cucumber.java.Scenario;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class Hooks {

    private static final ThreadLocal<Browser> browserThreadLocal = new ThreadLocal<>();
    private static final ThreadLocal<BrowserContext> contextThreadLocal = new ThreadLocal<>();
    private static final ThreadLocal<Page> pageThreadLocal = new ThreadLocal<>();
    private static final ThreadLocal<Scenario> scenarioThreadLocal = new ThreadLocal<>();
    private static final ThreadLocal<PageCommonMethods> pageCommonMethodsThreadLocal = new ThreadLocal<>();

    private static final Logger logger = LoggerFactory.getLogger(Hooks.class);

    private static boolean isLastScenarioFeature = false;
    private static boolean isFirstScenarioInFeature = true;

    @Before
    public void setUp(Scenario scenario) {
        scenarioThreadLocal.set(scenario);

        if (scenario.getSourceTagNames().contains("@LastScenario")) {
            isLastScenarioFeature = true;
        }

        if (isLastScenarioFeature && !isFirstScenarioInFeature) {
            logger.info("Reusing browser instance for feature with @LastScenario tag.");
            return;
        }

        try {
            String browserName = ConfigurationProperties.getBrowser();
            BrowserFactory.BrowserTypeEnum browserTypeEnum = switch (browserName.toUpperCase()) {
                case "CHROME" -> BrowserFactory.BrowserTypeEnum.CHROME;
                case "FIREFOX" -> BrowserFactory.BrowserTypeEnum.FIREFOX;
                case "WEBKIT" -> BrowserFactory.BrowserTypeEnum.WEBKIT;
                case "EDGE" -> BrowserFactory.BrowserTypeEnum.EDGE;
                default -> throw new IllegalArgumentException("Unsupported browser type: " + browserName);
            };

            Browser browser = BrowserFactory.createBrowser(browserTypeEnum);
            browserThreadLocal.set(browser);

            BrowserContext context = BrowserFactory.createContextWithVideo(browser);
            contextThreadLocal.set(context);

            Page page = context.newPage();
            pageThreadLocal.set(page);

            PageCommonMethods pageCommonMethods = new PageCommonMethods(page);
            pageCommonMethodsThreadLocal.set(pageCommonMethods);

            logger.info("Browser setup completed for scenario: {}", scenario.getName());
        } catch (Exception e) {
            logger.error("Error setting up the browser for scenario: {}", e.getMessage());
            throw new RuntimeException("Browser setup failed", e);
        }
    }

    @After
    public void tearDown(Scenario scenario) {
        try {
            if (scenario.isFailed()) {
                // Your failure handling logic can go here, e.g., taking screenshots.
            }
        } catch (Exception e) {
            logger.error("Error during scenario teardown: {}", e.getMessage(), e);
        } finally {

            ApiRequestHandler.disposeContext();

            // Safely close the database connection for the current thread.
            DatabaseHandler.closeConnection();

            // Your existing browser teardown logic follows.
            if (isLastScenarioFeature) {
                logger.info("Skipping browser closure for feature with @LastScenario tag.");
                isFirstScenarioInFeature = false;
            } else {
                closeBrowserResources();
            }
        }
    }

    private void closeBrowserResources() {
        try {
            Page page = pageThreadLocal.get();
            if (page != null && !page.isClosed()) {
                page.close();
            }
        } catch (Exception e) {
            logger.error("Error closing the page: {}", e.getMessage(), e);
        } finally {
            pageThreadLocal.remove();
        }

        try {
            BrowserContext context = contextThreadLocal.get();
            if (context != null) {
                context.close();
            }
        } catch (Exception e) {
            logger.error("Error closing the browser context: {}", e.getMessage(), e);
        } finally {
            contextThreadLocal.remove();
        }

        try {
            Browser browser = browserThreadLocal.get();
            if (browser != null) {
                browser.close();
                logger.info("Browser closed.");
            }
        } catch (Exception e) {
            logger.error("Error closing the browser: {}", e.getMessage(), e);
        } finally {
            browserThreadLocal.remove();
        }
    }

    public static Page getPage() {
        Page page = pageThreadLocal.get();
        if (page == null || page.isClosed()) {
            throw new IllegalStateException("The page is closed or not initialized.");
        }
        return page;
    }

    public static Browser getBrowser() {
        Browser browser = browserThreadLocal.get();
        if (browser == null) {
            throw new IllegalStateException("The browser is not initialized.");
        }
        return browser;
    }

    public static Scenario getCurrentScenario() {
        return scenarioThreadLocal.get();
    }

    public static void setCurrentScenario(Scenario scenario) {
        scenarioThreadLocal.set(scenario);
    }
}

File: ptaf/hooks.py
"""Runtime hook context replacing Java Hooks ThreadLocal accessors."""

from __future__ import annotations

from contextvars import ContextVar
from typing import Any

from playwright.sync_api import Browser, Page

from ptaf.ui.page_common import PageCommonMethods

_playwright: ContextVar[Any | None] = ContextVar("_playwright", default=None)
_browser: ContextVar[Browser | None] = ContextVar("_browser", default=None)
_context: ContextVar[Any | None] = ContextVar("_context", default=None)
_page: ContextVar[Page | None] = ContextVar("_page", default=None)
_scenario: ContextVar[Any | None] = ContextVar("_scenario", default=None)
_page_common: ContextVar[PageCommonMethods | None] = ContextVar(
    "_page_common", default=None
)


def set_playwright(playwright: Any | None) -> None:
    _playwright.set(playwright)


def get_playwright() -> Any | None:
    return _playwright.get()


def set_browser(browser: Browser | None) -> None:
    _browser.set(browser)


def get_browser() -> Browser:
    browser = _browser.get()
    if browser is None:
        raise RuntimeError("The browser is not initialized.")
    return browser


def set_context(context: Any | None) -> None:
    _context.set(context)


def get_context() -> Any | None:
    return _context.get()


def set_page(page: Page | None) -> None:
    _page.set(page)


def get_page() -> Page:
    page = _page.get()
    if page is None or page.is_closed():
        raise RuntimeError("The page is closed or not initialized.")
    return page


def set_current_scenario(scenario: Any | None) -> None:
    _scenario.set(scenario)
    PageCommonMethods.set_current_scenario(scenario)


def get_current_scenario() -> Any | None:
    return _scenario.get()


def set_page_common(page_common: PageCommonMethods | None) -> None:
    _page_common.set(page_common)


def get_page_common() -> PageCommonMethods | None:
    return _page_common.get()


def clear_runtime_context() -> None:
    _playwright.set(None)
    _browser.set(None)
    _context.set(None)
    _page.set(None)
    _scenario.set(None)
    _page_common.set(None)
